File: Back-End/src/main/java/org/sid/backend/service/EventService.java
package org.sid.backend.service;

import lombok.RequiredArgsConstructor;
import org.sid.backend.model.Event;
import org.sid.backend.model.Review;
import org.sid.backend.model.User;
import org.sid.backend.repository.EventRepository;
import org.sid.backend.repository.ReviewRepository;
import org.sid.backend.repository.UserRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.Date;
import java.util.List;
import java.util.Optional;

@Service
@RequiredArgsConstructor
public class EventService {

    private final UserRepository userRepository;
    private final EventRepository eventRepository;
    private final ReviewRepository reviewRepository;

    // Retrieve all events
    public List<Event> getAllEvents() {
        return eventRepository.findAll();
    }

    // Create an event with an organizer
    @Transactional
    public Event createEvent(Event event) {
        String organizerId = event.getOrganizerId(); // Assume organizerId is passed in the event object.

        if (organizerId == null || organizerId.isEmpty()) {
            throw new IllegalArgumentException("Organizer ID cannot be null or empty.");
        }

        // Fetch the organizer from the database
        User organizer = userRepository.findById(organizerId)
                .orElseThrow(() -> new IllegalArgumentException("Organizer not found with ID: " + organizerId));

        // Associate the organizer with the event
        event.setOrganizer(organizer);

        return eventRepository.save(event);
    }

    // Search events by title
    public List<Event> searchEventsByTitle(String title) {
        return eventRepository.findByTitleContainingIgnoreCase(title);
    }

    // Delete an event by ID
    @Transactional
    public void deleteEvent(String eventId) {
        if (!eventRepository.existsById(eventId)) {
            throw new IllegalArgumentException("Event not found with ID: " + eventId);
        }
        eventRepository.deleteById(eventId);
    }

    // Retrieve events by organizer ID
    public List<Event> getEventsByOrganizer(String organizerId) {
        User organizer = userRepository.findById(organizerId)
                .orElseThrow(() -> new IllegalArgumentException("Organizer not found with ID: " + organizerId));
        return eventRepository.findByOrganizer(organizer);
    }

    // Retrieve events by organizer email
    public List<Event> getEventsByOrganizerEmail(String email) {
        User organizer = userRepository.findByEmail(email)
                .orElseThrow(() -> new IllegalArgumentException("Organizer not found with email: " + email));
        return eventRepository.findByOrganizer(organizer);
    }

    // Retrieve events by category
    public List<Event> getEventsByCategory(String category) {
        return eventRepository.findByCategoryIgnoreCase(category);
    }

    // Retrieve events by location
    public List<Event> searchEventsByLocation(String location) {
        return eventRepository.findByLocationContainingIgnoreCase(location);
    }

    // Retrieve top-rated events
    public List<Event> getTopRatedEvents(double rating) {
        return eventRepository.findByRatingGreaterThanEqual(rating);
    }

    // Retrieve events by exact date
    public List<Event> getEventsByDate(Date date) {
        return eventRepository.findByDate(date);
    }

    // Retrieve upcoming events
    public List<Event> getUpcomingEvents(Date currentDate) {
        return eventRepository.findByDateAfter(currentDate);
    }

    // Retrieve past events
    public List<Event> getPastEvents(Date currentDate) {
        return eventRepository.findByDateBefore(currentDate);
    }

    // Retrieve events by category and rating
    public List<Event> getEventsByCategoryAndRating(String category, double minRating) {
        return eventRepository.findByCategoryAndRatingGreaterThanEqual(category, minRating);
    }

    // Retrieve an event by ID
    public Event getEventById(String id) {
        return eventRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Event not found with ID: " + id));
    }

    // Retrieve events by location and organizer
    public List<Event> getEventsByLocationAndOrganizer(String location, String organizerId) {
        User organizer = userRepository.findById(organizerId)
                .orElseThrow(() -> new IllegalArgumentException("Organizer not found with ID: " + organizerId));
        return eventRepository.findByLocationAndOrganizer(location, organizer);
    }

    // Update an existing event
    @Transactional
    public Event updateEvent(String eventId, Event updatedEvent) {
        Event existingEvent = eventRepository.findById(eventId)
                .orElseThrow(() -> new IllegalArgumentException("Event not found with ID: " + eventId));

        // Update fields if they are not null or zero
        if (updatedEvent.getTitle() != null) existingEvent.setTitle(updatedEvent.getTitle());
        if (updatedEvent.getDescription() != null) existingEvent.setDescription(updatedEvent.getDescription());
        if (updatedEvent.getDate() != null) existingEvent.setDate(updatedEvent.getDate());
        if (updatedEvent.getTime() != null) existingEvent.setTime(updatedEvent.getTime());
        if (updatedEvent.getLocation() != null) existingEvent.setLocation(updatedEvent.getLocation());
        if (updatedEvent.getCategory() != null) existingEvent.setCategory(updatedEvent.getCategory());
        if (updatedEvent.getRating() > 0) existingEvent.setRating(updatedEvent.getRating());
        if (updatedEvent.getVenue() != null) existingEvent.setVenue(updatedEvent.getVenue());
        if (updatedEvent.getCapacity() > 0) existingEvent.setCapacity(updatedEvent.getCapacity());

        return eventRepository.save(existingEvent);
    }

    public void updateEventRating(String eventId) {
        List<Review> reviews = reviewRepository.findByEventId(eventId);
        double averageRating = reviews.stream()
                .mapToInt(Review::getRating)
                .average()
                .orElse(0.0);

        Event event = eventRepository.findById(eventId)
                .orElseThrow(() -> new IllegalArgumentException("Event not found with ID: " + eventId));
        event.setRating(averageRating);
        eventRepository.save(event);
    }






    public double submitRating(String eventId, String userId, int rating) {
        // Validate rating range
        if (rating < 1 || rating > 5) {
            throw new IllegalArgumentException("Rating must be between 1 and 5.");
        }

        // Retrieve the event
        Event event = eventRepository.findById(eventId)
                .orElseThrow(() -> new IllegalArgumentException("Event not found with ID: " + eventId));

        // Check if the user has already rated
        Optional<Review> optionalReview = reviewRepository.findByEventIdAndUserId(eventId, userId);
        if (optionalReview.isPresent()) {
            // User has already rated, update the existing review
            Review existingReview = optionalReview.get();
            System.out.println("User has already rated. Updating their rating...");
            int previousRating = existingReview.getRating();
            existingReview.setRating(rating);
            reviewRepository.save(existingReview);

            // Calculate the new average rating
            double totalRating = (event.getRating() * event.getNumberOfRatings()) - previousRating + rating;
            double newAverageRating = totalRating / event.getNumberOfRatings();
            event.setRating(newAverageRating);
            eventRepository.save(event);

            return newAverageRating;
        } else {
            // User has not rated yet, create a new review
            System.out.println("New rating from user.");
            Review newReview = new Review();
            newReview.setEventId(eventId);
            newReview.setUserId(userId);
            newReview.setRating(rating);
            reviewRepository.save(newReview);

            // Calculate the new average rating
            double totalRating = (event.getRating() * event.getNumberOfRatings()) + rating;
            int newNumberOfRatings = event.getNumberOfRatings() + 1;
            double newAverageRating = totalRating / newNumberOfRatings;
            event.setRating(newAverageRating);
            event.setNumberOfRatings(newNumberOfRatings);
            eventRepository.save(event);

            return newAverageRating;
        }
    }

}

File: Front-ang-app/src/app/services/event.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { catchError, map, Observable, throwError } from 'rxjs';
import { AppEvent, Review } from '../models/event';

@Injectable({
  providedIn: 'root',
})
export class EventService {
  private apiUrl = '/api/events';

  constructor(private http: HttpClient) {}

  // Retrieve all events
  getAllEvents(): Observable<AppEvent[]> {
    return this.http.get<any[]>(this.apiUrl).pipe(
      map((data) =>
        data.map((item) => ({
          id: item.id,
          title: item.title,
          description: item.description,
          date: item.date,
          time: item.time,
          image: item.image,
          rating: item.rating,
          venue: item.venue,
          category: item.category,
          capacity: item.capacity, // Fixed field name
          location: item.location,
          organizerId: item.organizerId,
          organiser: item.organiser
            ? {
              id: item.organiser.id,
              name: item.organiser.name,
              email: item.organiser.email,
              role: item.organiser.role,
              password: item.organiser.password || '', // Optional field
            }
            : undefined,
          reviews: item.reviews || [],
        }))
      ),
      catchError((error) => {
        console.error('Error fetching events:', error);
        return throwError(() => error);
      })
    );
  }

  // Create a new event
  createEvent(event: AppEvent): Observable<AppEvent> {
    return this.http.post<AppEvent>(this.apiUrl, event).pipe(
      catchError((error) => {
        console.error('Error creating event:', error);
        return throwError(() => error);
      })
    );
  }

  // Retrieve event by ID
  // getEventById(id: string): Observable<AppEvent> {
  //   return this.http.get<AppEvent>(`${this.apiUrl}/${id}`).pipe(
  //     catchError((error) => {
  //       console.error('Error fetching event by ID:', error);
  //       return throwError(() => error);
  //     })
  //   );
  // }

  // Search events by title
  searchEventsByTitle(title: string): Observable<AppEvent[]> {
    const params = new HttpParams().set('title', title);
    return this.http.get<AppEvent[]>(`${this.apiUrl}/search`, { params }).pipe(
      catchError((error) => {
        console.error('Error searching events by title:', error);
        return throwError(() => error);
      })
    );
  }

  // Delete an event
  deleteEvent(id: string): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`).pipe(
      catchError((error) => {
        console.error('Error deleting event:', error);
        return throwError(() => error);
      })
    );
  }

  // Retrieve events by organizer
  getEventsByOrganizer(organizerId: string): Observable<AppEvent[]> {
    return this.http.get<AppEvent[]>(`${this.apiUrl}/organizer/${organizerId}`).pipe(
      catchError((error) => {
        console.error('Error fetching events by organizer:', error);
        return throwError(() => error);
      })
    );
  }

  // Retrieve events by category
  getEventsByCategory(category: string): Observable<AppEvent[]> {
    return this.http.get<AppEvent[]>(`${this.apiUrl}/category/${category}`).pipe(
      catchError((error) => {
        console.error('Error fetching events by category:', error);
        return throwError(() => error);
      })
    );
  }

  // Retrieve events by date
  getEventsByDate(date: string): Observable<AppEvent[]> {
    const params = new HttpParams().set('date', date);
    return this.http.get<AppEvent[]>(`${this.apiUrl}/date`, { params }).pipe(
      catchError((error) => {
        console.error('Error fetching events by date:', error);
        return throwError(() => error);
      })
    );
  }

  // Retrieve upcoming events
  getUpcomingEvents(currentDate: string): Observable<AppEvent[]> {
    const params = new HttpParams().set('currentDate', currentDate);
    return this.http.get<AppEvent[]>(`${this.apiUrl}/upcoming`, { params }).pipe(
      catchError((error) => {
        console.error('Error fetching upcoming events:', error);
        return throwError(() => error);
      })
    );
  }

  // Retrieve past events
  getPastEvents(currentDate: string): Observable<AppEvent[]> {
    const params = new HttpParams().set('currentDate', currentDate);
    return this.http.get<AppEvent[]>(`${this.apiUrl}/past`, { params }).pipe(
      catchError((error) => {
        console.error('Error fetching past events:', error);
        return throwError(() => error);
      })
    );
  }

  // Retrieve events by category and rating
  getEventsByCategoryAndRating(category: string, minRating: number): Observable<AppEvent[]> {
    const params = new HttpParams()
      .set('category', category)
      .set('minRating', minRating.toString());
    return this.http.get<AppEvent[]>(`${this.apiUrl}/category-rating`, { params }).pipe(
      catchError((error) => {
        console.error('Error fetching events by category and rating:', error);
        return throwError(() => error);
      })
    );
  }

  // Retrieve events by location
  searchEventsByLocation(location: string): Observable<AppEvent[]> {
    const params = new HttpParams().set('location', location);
    return this.http.get<AppEvent[]>(`${this.apiUrl}/location`, { params }).pipe(
      catchError((error) => {
        console.error('Error searching events by location:', error);
        return throwError(() => error);
      })
    );
  }

  // Retrieve events by location and organizer
  getEventsByLocationAndOrganizer(location: string, organizerId: string): Observable<AppEvent[]> {
    const params = new HttpParams()
      .set('location', location)
      .set('organizerId', organizerId);
    return this.http.get<AppEvent[]>(`${this.apiUrl}/location-organizer`, { params }).pipe(
      catchError((error) => {
        console.error('Error fetching events by location and organizer:', error);
        return throwError(() => error);
      })
    );
  }

  // Retrieve top-rated events
  getTopRatedEvents(rating: number): Observable<AppEvent[]> {
    const params = new HttpParams().set('rating', rating.toString());
    return this.http.get<AppEvent[]>(`${this.apiUrl}/rating`, { params }).pipe(
      catchError((error) => {
        console.error('Error fetching top-rated events:', error);
        return throwError(() => error);
      })
    );
  }

  // Update an existing event
  updateEvent(id: string, updatedEvent: AppEvent): Observable<AppEvent> {
    return this.http.put<AppEvent>(`${this.apiUrl}/${id}`, updatedEvent).pipe(
      catchError((error) => {
        console.error('Error updating event:', error);
        return throwError(() => error);
      })
    );
  }

  // Retrieve events by organizer's email
  getEventsByOrganizerEmail(email: string): Observable<AppEvent[]> {
    const params = new HttpParams().set('email', email);
    return this.http.get<AppEvent[]>(`${this.apiUrl}/organizer/email`, { params }).pipe(
      catchError((error) => {
        console.error('Error fetching events by organizer email:', error);
        return throwError(() => error);
      })
    );
  }
  getEventById(id: string): Observable<AppEvent> {
    return this.http.get<any>(`${this.apiUrl}/${id}`).pipe(
      map((data) => ({
        id: data.id,
        title: data.title,
        description: data.description,
        date: data.date,
        time: data.time,
        image: data.image,
        rating: data.rating,
        venue: data.venue,
        category: data.category,
        capacity: data.capacity, // Correct field name
        location: data.location,
        organizerId: data.organizerId,
        organiser: data.organiser
          ? {
            id: data.organiser.id,
            name: data.organiser.name,
            email: data.organiser.email,
            role: data.organiser.role,
            password: data.organiser.password || '', // Optional handling
          }
          : undefined,
        reviews: data.reviews || [],
      })),
      catchError((error) => {
        console.error('Error fetching event by ID:', error);
        return throwError(() => error);
      })
    );
  }

  // Submit rating for an event

  submitEventRating(url: string, data: any): Observable<any> {
    return this.http.post<any>(url, data).pipe(
      catchError((error) => {
        console.error('Error in submitEventRating:', error);
        return throwError(() => error);
      })
    );
  }







}
